bin/tx-prover/src/proxy/mod.rs

```diff
@@ -0,0 +1,280 @@
+
+        // Read request ID from headers
+        let request_id = Self::get_request_id(session)?;
+
+        // Enqueue the request ID in the worker queue
+        // We use a new scope to release the lock after the operation
+        {
+            let mut ctx_guard = QUEUES.write().await;
+            let worker_queue = ctx_guard.entry(worker.clone()).or_insert_with(Vec::new);
+
+            // Limit queue length to MAX_QUEUE_ITEMS requests
+            if worker_queue.len() >= MAX_QUEUE_ITEMS {
+                return Err(Error::new_str("Too many requests in the queue"));
+            }
+
+            worker_queue.push(request_id.to_string());
+        }
+
+        // Wait for the request to be at the front of the queue
+        loop {
+            // We use a new scope for each iteration to release the lock
+            {
+                let ctx_guard = QUEUES.read().await;
+                if let Some(worker_queue) = ctx_guard.get(&worker) {
+                    if worker_queue[0] == request_id {
+                        break;
+                    }
+                } else {
+                    return Err(Error::new_str("Worker not found"));
+                }
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+
+        // Set SNI
+        let mut http_peer = HttpPeer::new(worker, false, "".to_string());
+        let peer_opts =
+            http_peer.get_mut_peer_options().ok_or(Error::new(ErrorType::InternalError))?;
+
+        // Timeout settings
+        peer_opts.total_connection_timeout = TIMEOUT_SECS;
+        peer_opts.connection_timeout = CONNECTION_TIMEOUT_SECS;
+        peer_opts.read_timeout = TIMEOUT_SECS;
+        peer_opts.write_timeout = TIMEOUT_SECS;
+        peer_opts.idle_timeout = TIMEOUT_SECS;
+
+        // Enable HTTP/2
+        peer_opts.alpn = ALPN::H2;
+
+        let peer = Box::new(http_peer);
+        Ok(peer)
+    }
+
+    /// Applies the necessary filters to the request before sending it to the upstream server.
+    ///
+    /// Here we ensure that the correct headers are forwarded for gRPC requests.
+    ///
+    /// This method is called right after [Self::upstream_peer()] returns a [HttpPeer] and a
+    /// connection is established with the worker.
+    async fn upstream_request_filter(
+        &self,
+        _session: &mut Session,
+        upstream_request: &mut RequestHeader,
+        _ctx: &mut Self::CTX,
+    ) -> Result<()>
+    where
+        Self::CTX: Send + Sync,
+    {
+        // Check if it's a gRPC request
+        if let Some(content_type) = upstream_request.headers.get("content-type") {
+            if content_type == "application/grpc" {
+                // Ensure the correct host and gRPC headers are forwarded
+                upstream_request.insert_header("content-type", "application/grpc")?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Retry the request if the connection fails.
+    fn fail_to_connect(
```

Comment:
Question: does it use any delays between retry attempts? If not, should we add it? But either way, we should mention a bit more about how retry strategy works.

Reply:
It does not have any delay in case of a retry. I'm not sure if it needs it since we are supposed to be using a different worker for the retry. Ok, will mention this changes in the existing issues to also attack this

Reply:
Ah - interesting! One thing that is not immediately clear to me is how it would pick a different worker. For this to happen, the request needs to be removed from the queue of one worker and added to the queue of a different worker, right? Would this happen naturally (e.g., would the request get a different ID or something like that)?

Reply:
It does need to be removed from the queue, that's why the first part of `upstream_peer` [checks if it is a retry](https://github.com/0xPolygonMiden/miden-base/pull/930/files#diff-dbbcd3a5ea499f3ce31eb5c3398931ee02004a6cb109787b150bcb5e7d9c6555R158-R160), and in that case removes the request from the respective worker queue. I'm testing a couple thing to change this for a better handled retry case, moving the dequeueing to the `fail_to_connect` method and using `tokio::Notify` instead of a sleep to await for the turn of a request to be processed.

Reply:
I opened an issue: https://github.com/0xPolygonMiden/miden-base/issues/947